singular_locus for projective schemes

paemurru opened the pull request:
Pinging @afkafkafk13

paemurru committed:
singular_locus for projective schemes

paemurru committed:
Fix documentation

afkafkafk13 commented:
Here I object to the way of including this, not to the idea to have this. 

We have put a lot of energy into the affine case for handling it in all generality and this carries over to the projective case in the following way:
1) compute the singular locus of the affine cone
2) make sure to ignore the vertex of the affine cone
3) interpret the result as equations of the projective subvariety defining the singular locus of your projective variety

Translated to Oscar this means:
1) Forget the grading
2) compute the singular locus 
3) saturate away the vertex of the affine cone
4) move result back to the graded case and define the subvariety

In this way, we do not need to restrict ourselves to the equidimensional case and at the same time we do not need a full primary decomposition. If you need help to guide you through the different options and methods in the affine case, let me know.

afkafkafk13 commented:
P.S.: In this way we avoid code duplication which is always a source of mysterious bugs and pitfalls, as soon as the duplicate fragments start to evolve independently.

HechtiDerLachs commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -138,6 +138,53 @@ function _projective_jacobian_criterion(P::AbsProjectiveScheme{<:Ring, <:MPolyQu
+      :is_smooth,
+      "projective Jacobian criterion only implemented when scheme is equidimensional. Instead, you could try `singular_locus(covered_scheme(X))`"
+    ))
+  end
```
```suggestion
function singular_locus(P::AbsProjectiveScheme{<:Ring, <:MPolyQuoRing}; check::Bool=true)
  @check begin
    if !(is_equidimensional(P))
      throw(NotImplementedError(
        :is_smooth,
        "projective Jacobian criterion only implemented when scheme is equidimensional. Instead, you could try `singular_locus(covered_scheme(X))`"
      ))
    end
  end
```
The equidimensionality-check is tentatively expensive, so we probably want a way to turn it off. This should do the job.

afkafkafk13 replied:
Indeed..

HechtiDerLachs commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -138,6 +138,53 @@ function _projective_jacobian_criterion(P::AbsProjectiveScheme{<:Ring, <:MPolyQu
+  mat = jacobian_matrix(R, gens(I))
+  sing_locus = ideal(R, minors(mat, codim(P))) + I
+  irrelevant_ideal = ideal(R, gens(R))
+  sing_locus_saturated = saturation(sing_locus, irrelevant_ideal)
```
@wdecker once suggested a quicker way to do this. Since everything is homogeneous, it suffices to compute the dimension of the ideal `sing_locus`: If it is zero, then it must be primary to the irrelevant ideal.

afkafkafk13 replied:
Good Point! Thx

simonbrandhorst replied:
We could just leave out this line (I do not know how expensive it is).
I am not sure how knowing the dimension helps here. Since we want the singular locus and not its dimension.

afkafkafk13 replied:
If the (affine) dimension is higher than zero, then we return the locus as is (it is correct, but might contain powers of the irrelevant ideal).  If it is zero, return empty set without saturation.

HechtiDerLachs reviewed (commented):
I think this is a reasonable contribution to the code base. 

I agree with @afkafkafk13 that we probably want to have some more ways to check smoothness. But in my opinion, these can be made available via keyword arguments also later. Hence, I do not see any reason not to go forward with this for the moment.

HechtiDerLachs commented:
~~One~~ Edit: two more remarks: 
 * Some other methods have the keyword argument `compute_radical`. Maybe we want to include that here, too? If it is set to `true` then the radical of the singular locus is computed (which makes it unique).
 * The current method is for projective schemes with quotient rings. We should add a dummy method for the other case, too (which just returns `true`). However, we probably want to restrict to the case where the coefficient ring is a field? That should be the first type parameter of `AbsProjectiveScheme`.

afkafkafk13 commented:
Still, there is more to the implementation of the jacobian criterion in the affine case and we should use this as a backend. What we need here is just an appropriate interface to it, not a new implementation of the jacobian criterion.

(@HechtiDerLachs : I am not talking about is_smooth, but about singular_locus -- which of course internally uses the jacobian criterion)

simonbrandhorst commented:
> Here I object to the way of including this, not to the idea to have this.
> 
> We have put a lot of energy into the affine case for handling it in all generality and this carries over to the projective case in the following way:
> 
>     1. compute the singular locus of the affine cone
> 
>     2. make sure to ignore the vertex of the affine cone
> 
>     3. interpret the result as equations of the projective subvariety defining the singular locus of your projective variety
> 
> 
> Translated to Oscar this means:
> 
>     1. Forget the grading
> 
>     2. compute the singular locus
> 
>     3. saturate away the vertex of the affine cone
> 
>     4. move result back to the graded case and define the subvariety
> 
> 
> In this way, we do not need to restrict ourselves to the equidimensional case and at the same time we do not need a full primary decomposition. If you need help to guide you through the different options and methods in the affine case, let me know.

Indeed, we have put a lot of effort in the `is_smooth` function. We should mimick its behavior here. 
One point is that the method working in utmost generality is not necessarily the fastest:
In the algorithm selection for `is_smooth` we do:
```
 if algorithm == :default
    if is_equidimensional(P)
      algorithm = :projective_jacobian
    elseif base_ring(P) isa Field
      algorithm = :affine_cone
``` 
and I seem to recall that we did some benchmarking at the time.  So if you want to make the affine cone algorithm a standard. That would require some benchmarking to provide sound evidence that it is generically faster. 

Also we do not need to restrict to fields. Being smooth/singular is relative over the base and we do the same for `is_smooth` .

But I agree that if we are not equidimensional, then can fall back to the affine cone method.

afkafkafk13 commented:
> Indeed, we have put a lot of effort in the `is_smooth` function. We should mimick its behavior here. One point is that the method working in utmost generality is not necessarily the fastest: In the algorithm selection for `is_smooth` we do:
> 
> ```
>  if algorithm == :default
>     if is_equidimensional(P)
>       algorithm = :projective_jacobian
>     elseif base_ring(P) isa Field
>       algorithm = :affine_cone
> ```
> 
> and I seem to recall that we did some benchmarking at the time. So if you want to make the affine cone algorithm a standard. That would require some benchmarking to provide sound evidence that it is generically faster.
> 
> Also we do not need to restrict to fields. Being smooth/singular is relative over the base and we do the same for `is_smooth` .
> 
> But I agree that if we are not equidimensional, then can fall back to the affine cone method.

Then our positions are not far from each other: If there are special settings with faster algorithms, we should  of course use them (and benefit from the experience in the affine case), but instead of just throwing an error for more general situations, there is still the affine cone fallback. By the way, I am not talking about `is_smooth`, but about `singular_locus`. For `is_smooth`we have more room to come to the result than for `singular_locus`.

paemurru committed:
Avoid code duplication, add tests

paemurru commented:
* Similarly to `is_smooth`, we use the algorithm `:projective_jacobian` if the scheme is equidimensional and the algorithm `:affine_cone` is the coefficient ring is a field. Note that `singular_locus` does not work for affine schemes when the coefficient ring is not a field, so this is the best we can do for now.
* Added keyword argument `saturate` to set whether we saturate or not.
* Added tests.
* Added the case of projective spaces.

simonbrandhorst reviewed (approved):
Looks good from my point of view. This is certainly an improvement. 
If someone wants to improve this even further, they may open a pull request themselves.

paemurru committed:
Add `compute_radical`

paemurru committed:
Rename `is_saturate` to `saturate`

paemurru committed:
Describe keyword arguments

Also add algorithms for projective spaces.

afkafkafk13 commented:
Can you give Erik and me until monday before merging?

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -60,7 +60,7 @@ false
 is_smooth(P::AbsProjectiveScheme; algorithm::Symbol=:default)
 
-@attr Bool function is_smooth(P::AbsProjectiveScheme{<:Any, <:MPolyQuoRing}; algorithm::Symbol=:default)
+@attr Bool function is_smooth(P::AbsProjectiveScheme{<:Ring, <:MPolyQuoRing}; algorithm::Symbol=:default)
```
For lack of possibility to comment to line 69 right below, I comment here:
`is_equidimensional` passes through a (usually expensive) equidimensional decomposition, which is thrown away afterwards and then recalculated in the non-equidimensional case. We should avoid this. 

Better alternative: check, `has_attribute(X, :is_equidimensional) && is_equidimensional(X)` and then pass to the affine cone if the check returns false
(@simonbrandhorst @HechtiDerLachs : This should actually be done by the is_known mechanism, but I do not have a good example to point Erik to)

It would actually make sense to check is_irreducible before is_equidimensional, as irreducible imples equidimensional and should/might be set in many standard settings (e.g. projective varieties). 

Sideremark: `is_equidimensional` for ideals in polynomial rings should not be located in the IdealSheaves subdirectory of the Scheme framework, but in the respective files in src/Rings .

wdecker replied:
See my remark in Pr #5066. I agree with your sideremark.

afkafkafk13 replied:
I completely agree with you on the equidimensional hull statements made there.

On the other hand, here I am inclined to only test for known properties on this level and then leave the equidimensional decomposition to the affine method, which really needs it and uses the data.

simonbrandhorst replied:
Then `is_irreducible` should be checked in the `is_equidimensional` function. And this should be done in another PR.

simonbrandhorst replied:
By the way, the `equidimensional_decomposition_weak` is cached by default.
```julia> @time equidimensional_decomposition_weak(I)
  1.047411 seconds (21.64 k allocations: 784.156 KiB)
1-element Vector{MPolyIdeal{MPolyDecRingElem{FqFieldElem, FqMPolyRingElem}}}:
 Ideal with 6 generators

julia> @time equidimensional_decomposition_weak(I)
  0.000006 seconds
1-element Vector{MPolyIdeal{MPolyDecRingElem{FqFieldElem, FqMPolyRingElem}}}:
 Ideal with 6 generators
```
so it is not thrown away.

simonbrandhorst replied:
Can you give timings that the `:affine_cone` method is faster than `:projective_jacobian` (including a check for equidimensionality?)

afkafkafk13 replied:
Two remarks, since I am not primarily aiming at timings:
1) projective_jacobian and the affine_cone approach do exactly the same thing with the exception that projective_jacobian is not implemented for the general case
2) in the affine_cone approach, the affine cone is a new object, which does not profit from the caching of the equidimensional decomposition of the projective scheme

simonbrandhorst replied:
Okay, I see, they do the same thing. Then I wonder why we do things as we do for `is_smooth`, 
which besides is buggy:
```
julia> R,(x,y) = polynomial_ring(GF(3),2)
(Multivariate polynomial ring in 2 variables over GF(3), FqMPolyRingElem[x1, x2])

julia> I = ideal(x^3+y^3)
Ideal generated by
  x1^3 + x2^3

julia> is_smooth(spec(I))
true

```
(The tangent bundle is locally free, but it has the wrong dimension.)

afkafkafk13 replied:
I agree. And I am wondering the same thing.... (historical reasons -- unreported or unresolved issue?)

This is_smooth is neither the jacobian criterion nor the is_regular approach, which originates from Janko and me.

afkafkafk13 replied:
Further remark on is_smooth: We need to be careful about smooth and regular, which differ in the general case.

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -102,28 +102,33 @@ is_smooth(P::AbsProjectiveScheme; algorithm::Symbol=:default)
+  aff, _ = affine_cone(P)
+  sing, _ = singular_locus(aff; compute_radical)
+  !saturate && return saturated_ideal(defining_ideal(sing))
+  origin = ideal(gens(ambient_coordinate_ring(sing)))
```
maybe check dim(defining_ideal(sing)) here between lines 126 adn 127

if dim(defining_ideal(sing)) is zero, the projective scheme is smooth and we can return the ideal generated by one in the graded ring without passing to saturation

Saturation can be expensive as it involved iterated groebner bases computations, whereas dim only requires a single one.

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -102,28 +102,33 @@ is_smooth(P::AbsProjectiveScheme; algorithm::Symbol=:default)
+  return saturation(saturated_ideal(defining_ideal(sing)), origin)
+end
+
+function _projective_jacobian_singular_locus_ideal(P::AbsProjectiveScheme{<:Ring, <:MPolyQuoRing}; saturate::Bool=true)
```
This is an internal function. You might just use `@req` for checking is_equidimensional || is_irreducible in line 132- 137. We should not end up here, unless we have ensured to at least have found an attribute certifying irreducibility or equidimensionality.

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -134,10 +139,105 @@ function _projective_jacobian_criterion(P::AbsProjectiveScheme{<:Ring, <:MPolyQu
   mat = jacobian_matrix(R, gens(I))
   sing_locus = ideal(R, minors(mat, codim(P))) + I
+  !saturate && return sing_locus
   irrelevant_ideal = ideal(R, gens(R))
-  return is_one(saturation(sing_locus, irrelevant_ideal))
+  return saturation(sing_locus, irrelevant_ideal)
```
as above: comment on dimension... for bypassing a saturation in the case of an empty singular locus.

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -134,10 +139,105 @@ function _projective_jacobian_criterion(P::AbsProjectiveScheme{<:Ring, <:MPolyQu
+  end
+
+  if algorithm == :default
+    if is_equidimensional(P)
```
same equidimensionality remark as for `is_smooth`

afkafkafk13 reviewed (commented):
I left you a few remarks for potential improvements.

Commit:
improvements for the affine_cone variant of singular_locus/is_smooth

Commit:
moved is_equidimensional to the canonical places

Commit:
fix typo

Commit:
clean-up is_equidimensional even more

wdecker commented on src/Rings/mpoly-graded.jl:
```diff
@@ -2578,6 +2578,10 @@ function minimal_generating_set(I::MPolyIdeal{<:MPolyDecRingElem})
+##################compatibility functions##########
+
+is_equidimensional(I::MPolyIdeal{<:MPolyDecRingElem}) = is_equidimensional(forget_grading(I))
+
```
Why is this needed?

afkafkafk13 replied:
You are completely right. This is not necessary. I remove it.

paemurru committed:
Fix typo

Commit:
remove unnecessary compatibility function

Commit:
Merge branch 'ep/projective_singular_locus' of github.com:paemurru/Oscar.jl into ep/projective_singular_locus

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -102,28 +102,40 @@ is_smooth(P::AbsProjectiveScheme; algorithm::Symbol=:default)
+    return dim(SL) < 1
   elseif algorithm == :projective_jacobian
-    return _projective_jacobian_criterion(P)
+    return is_one(_projective_jacobian_singular_locus_ideal(P))
```
```suggestion
    return dim(_projective_jacobian_singular_locus_ideal(P;saturate=false))<=0
```

afkafkafk13 replied:
This would work for an affine scheme, but not for a projective one. See below, but I see your point.

```
julia> R,(x,y) = graded_polynomial_ring(QQ,2)
(Graded multivariate polynomial ring in 2 variables over QQ, MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x1, x2])

julia> I = ideal(R,[x,y])
Ideal generated by
  x1
  x2

julia> X = projective_scheme(R,I)
Projective scheme
  over rational field
defined by ideal (x1, x2)

julia> dim(X)
-1
```

simonbrandhorst replied:
Since `P` is an ideal it should be o.k.?
```
julia> R,(x,y) = graded_polynomial_ring(QQ,2)
(Graded multivariate polynomial ring in 2 variables over QQ, MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x1, x2])

julia> I = ideal(R,[x,y])
Ideal generated by
  x1
  x2

julia> dim(I)
0
```

afkafkafk13 replied:
By the way, this should again be a dimension of `-inf` for the scheme.

Commit:
avoid saturation in projective case

Commit:
fix

afkafkafk13 commented:
I do not see the failure of the booktests in my courtyard.

The general implementation of is_smooth and is_regular is beyond the scope of this PR.

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -134,10 +146,105 @@ function _projective_jacobian_criterion(P::AbsProjectiveScheme{<:Ring, <:MPolyQu
+    singular_locus(P::AbsProjectiveScheme; algorithm::Symbol=:default, saturate::Bool=true) -> AbsProjectiveScheme
+
+Given an equidimensional projective scheme, returns the possibly
+nonreduced subscheme of projective space describing the singular locus.
```
```suggestion
Given a projective scheme, return the possibly
nonreduced subscheme of projective space describing the singular locus.
```

simonbrandhorst committed:
Update src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl

simonbrandhorst commented:
Thanks! Looks good to me.
